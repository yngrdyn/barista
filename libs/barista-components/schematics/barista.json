{
  "title": "Shipped schematics",
  "postid": "schematics",
  "public": true,
  "toc": true,
  "draft": true,
  "tags": ["component", "angular"],
  "navGroup": "angular-component",
  "category": "components",
  "layout": "page-angular.hbs"
}
