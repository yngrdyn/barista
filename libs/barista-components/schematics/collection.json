{
  "$schema": "../../node_modules/@angular-devkit/schematics/collection-schema.json",
  "schematics": {
    "ng-add": {
      "description": "Add and migrate Dynatrace Angular Components to v5.",
      "factory": "./src/ng-add/index",
      "schema": "./src/ng-add/schema.json"
    },
    "update-5.0.0": {
      "version": "5.0.0",
      "description": "Updates from the old @dynatrace/angular-components to the new @dynatrace/barista-components",
      "factory": "./src/ng-update/index#updateToV5"
    }
  }
}
