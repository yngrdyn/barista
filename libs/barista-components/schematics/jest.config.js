module.exports = {
  name: 'schematics',
  preset: '../../../jest.config.js',
  coverageDirectory: '../../../coverage/components/schematics',
  snapshotSerializers: ['jest-serializer-path'],
};
