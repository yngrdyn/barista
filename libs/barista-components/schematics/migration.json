{
  "$schema": "../../node_modules/@angular-devkit/schematics/collection-schema.json",
  "schematics": {
    "update-5.0.0": {
      "version": "5.0.0",
      "description": "Updates from the old @dynatrace/angular-components to the new @dynatrace/barista-components",
      "factory": "./src/ng-update/index#updateToV5"
    }
  }
}
