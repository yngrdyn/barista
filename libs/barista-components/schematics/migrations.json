{
  "$schema": "../../../node_modules/@angular-devkit/schematics/collection-schema.json",
  "schematics": {
    "update-7-0-0": {
      "version": "7.0.0",
      "description": "Updates to the Major Version 7.0.0",
      "factory": "./src/migrations/update-7-0-0/update-7-0-0"
    }
  }
}
