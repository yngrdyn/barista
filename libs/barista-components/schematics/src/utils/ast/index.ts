/**
 * @license
 * Copyright 2020 Dynatrace LLC
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

export * from './add-import-to-source-file';
export * from './create-import-declaration';
export * from './create-property-assignment';
export * from './create-string-literal';
export * from './find-nodes';
export * from './get-symbol-name';
export * from './has-export-modifier';
export * from './print-node';
export * from './update-ng-module-decorator-properties';
