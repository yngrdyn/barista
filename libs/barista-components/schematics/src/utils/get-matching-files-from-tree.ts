/**
 * @license
 * Copyright 2020 Dynatrace LLC
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Tree } from '@angular-devkit/schematics/src/tree/interface';

/**
 * Returns a set of matched files in a tree by a provided matcher function
 * @param tree The tree where we have to search for the files
 * @param matcher A function that gets a file path and check if the path matches
 */
export function getMatchingFilesFromTree(
  tree: Tree,
  matcher: (filePath: string) => boolean,
): Set<string> {
  const files = new Set<string>();
  tree.root.visit((file) => {
    if (matcher(file)) {
      files.add(file.toString());
    }
  });
  return files;
}
