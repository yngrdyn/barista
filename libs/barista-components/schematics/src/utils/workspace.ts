/**
 * @license
 * Copyright 2020 Dynatrace LLC
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Tree, Rule } from '@angular-devkit/schematics';
import { workspaces } from '@angular-devkit/core';
import { readFileFromTree } from './read-file-from-tree';

function createHost(tree: Tree): workspaces.WorkspaceHost {
  return {
    async readFile(path: string): Promise<string> {
      return readFileFromTree(tree, path);
    },
    async writeFile(path: string, data: string): Promise<void> {
      return tree.overwrite(path, data);
    },
    async isDirectory(path: string): Promise<boolean> {
      // approximate a directory check
      return !tree.exists(path) && tree.getDir(path).subfiles.length > 0;
    },
    async isFile(path: string): Promise<boolean> {
      return tree.exists(path);
    },
  };
}

/** Get the angular workspace out of the tree */
export async function getWorkspace(
  tree: Tree,
  path: string = '/',
): Promise<workspaces.WorkspaceDefinition> {
  const host = createHost(tree);
  const { workspace } = await workspaces.readWorkspace(path, host);
  return workspace;
}

export function updateWorkspace(
  updater: (
    workspace: workspaces.WorkspaceDefinition,
  ) => void | PromiseLike<void>,
): Rule;

export function updateWorkspace(
  workspace: workspaces.WorkspaceDefinition,
): Rule;

export function updateWorkspace(
  updaterOrWorkspace:
    | workspaces.WorkspaceDefinition
    | ((workspace: workspaces.WorkspaceDefinition) => void | PromiseLike<void>),
): Rule {
  return async (tree: Tree) => {
    const host = createHost(tree);

    if (typeof updaterOrWorkspace === 'function') {
      const { workspace } = await workspaces.readWorkspace('/', host);

      const result = updaterOrWorkspace(workspace);
      if (result !== undefined) {
        // tslint:disable-next-line: await-promise
        await result;
      }

      await workspaces.writeWorkspace(workspace, host);
    } else {
      await workspaces.writeWorkspace(updaterOrWorkspace, host);
    }
  };
}
