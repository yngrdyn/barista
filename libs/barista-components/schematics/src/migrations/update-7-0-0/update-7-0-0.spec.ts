/**
 * @license
 * Copyright 2020 Dynatrace LLC
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Tree } from '@angular-devkit/schematics';
import { SchematicTestRunner } from '@angular-devkit/schematics/testing';
import { readJsonInTree, serializeJson } from '@nrwl/workspace';
import { createEmptyWorkspace } from '@nrwl/workspace/testing';
import * as path from 'path';
import { readFileFromTree } from '../../utils';

describe('Update 7.0.0', () => {
  let initialTree: Tree;
  let schematicRunner: SchematicTestRunner;

  beforeEach(async () => {
    initialTree = createEmptyWorkspace(Tree.empty());

    schematicRunner = new SchematicTestRunner(
      '@dynatrace/barista-components/schematics',
      path.join(__dirname, '../../../migrations.json'),
    );

    initialTree.overwrite(
      'package.json',
      serializeJson({
        dependencies: {
          '@dynatrace/barista-components': '6.0.0',
          '@dynatrace/barista-fonts': '1.0.1',
          '@dynatrace/barista-icons': '3.6.0',
          highcharts: '6.2.0',
        },
        devDependencies: {
          '@types/highcharts': '5.0.36',
        },
      }),
    );
  });

  it('should update the dpendencies without highcharts', async () => {
    initialTree.overwrite(
      'package.json',
      serializeJson({
        dependencies: {},
      }),
    );
    const result = await schematicRunner
      .runSchematicAsync('update-7-0-0', {}, initialTree)
      .toPromise();

    const { dependencies } = readJsonInTree(result, '/package.json');
    expect(dependencies).toEqual({
      '@dynatrace/barista-components': '^7.0.0',
      '@dynatrace/barista-icons': '^5.0.0',
      'lodash-es': '^4.17.15',
    });
  });

  it('should update the dependencies with highcharts and remove the highcharts types', async () => {
    const result = await schematicRunner
      .runSchematicAsync('update-7-0-0', {}, initialTree)
      .toPromise();

    const { dependencies, devDependencies } = readJsonInTree(
      result,
      '/package.json',
    );
    expect(dependencies).toEqual({
      '@dynatrace/barista-components': '^7.0.0',
      '@dynatrace/barista-icons': '^5.0.0',
      '@dynatrace/barista-fonts': '1.0.1',
      highcharts: '^7.2.1',
      'lodash-es': '^4.17.15',
    });
    expect(devDependencies).toEqual({});
  });

  const templateReplacements = [
    {
      description: 'should run the migration for the aria-label-close',
      fileName: '/libs/feature-a/src/lib/component-a/component-a.html',
    },
    {
      description:
        'should replace the aria-label-close-button on the context dialog',
      fileName: '/libs/feature-a/src/lib/context-dialog.html',
    },
    {
      description:
        'should replace the aria-label to the native one on the breadcrumbs',
      fileName: 'libs/feature-a/src/lib/aria-label.html',
    },
    {
      description:
        'should migrate the breadcrumbs-item selector to a plain anchor',
      fileName: 'libs/feature-a/src/lib/breadcrumbs-item.html',
    },
    {
      description:
        'should migrate the table empty state to the empty state component',
      fileName: 'libs/feature-a/src/lib/table-empty-state-simple.html',
    },
    {
      description:
        'should migrate the table empty state to the empty state component with directive on the message',
      fileName: 'libs/feature-a/src/lib/table-empty-state.html',
    },
    {
      description: 'should not migrate i18n aria label attributes',
      fileName: 'libs/feature-a/src/lib/secondary-nav.html',
    },
    {
      description: 'should not migrate normal aria label attributes',
      fileName: 'libs/feature-a/src/lib/secondary-nav-aria-label.html',
    },
    {
      description: 'should migrate aria label bindings',
      fileName: 'libs/feature-a/src/lib/aria-label-binding.html',
    },
  ];

  describe('Template replacements', () => {
    templateReplacements.forEach(({ description, fileName }) => {
      it(description, async () => {
        initialTree = Tree.merge(initialTree, addHTMLFiles());
        const result = await schematicRunner
          .runSchematicAsync('update-7-0-0', {}, initialTree)
          .toPromise();
        const file = readFileFromTree(result, fileName);

        expect(file).toMatchSnapshot();
      });
    });
  });
});

function addHTMLFiles(): Tree {
  const tree = Tree.empty();
  tree.create(
    'libs/feature-a/src/lib/component-a/component-a.html',
    `
  <dt-chart [options]="options" [series]="series$ | async">
    <dt-chart-timestamp
      aria-label-close="Close the selection"
      (closed)="closed()"
      (valueChanges)="valueChanges($event)"
    ></dt-chart-timestamp>
  </dt-chart>

  `,
  );

  tree.create(
    'libs/feature-a/src/lib/context-dialog.html',
    `<dt-context-dialog #contextDialog
    aria-label="Show dashboard modification options"
    aria-label-close-button="Hide dashboard modification options"
    i18n-aria-label
>`,
  );

  tree.create(
    'libs/feature-a/src/lib/aria-label.html',
    `<div *pageHeader  dtTheme='blue'>
    <dt-breadcrumbs aria-label="Breadcrumbs navigation" i18n-aria-label>
      <a dtBreadcrumbsItem [navLink]="cmcDsService.getHome()" i18n>Home</a>
      <a dtBreadcrumbsItem [navLink]="cmcDsService.getDeploymentStatusMainPage()" i18n>Deployment status</a>
      <a dtBreadcrumbsItem [navLink]="cmcDsService.getActiveGateOverview()" i18n>ActiveGates</a>
    </dt-breadcrumbs>
  </div>`,
  );

  tree.create(
    'libs/feature-a/src/lib/breadcrumbs-item.html',
    `<dt-breadcrumbs aria-label="Breadcrumbs navigation" i18n-aria-label>
  <dt-breadcrumbs-item i18n>Security</dt-breadcrumbs-item>
  <dt-breadcrumbs-item href="/home" i18n>Security 2</dt-breadcrumbs-item>
  <dt-breadcrumbs-item [href]="'/home'" i18n>Security 3</dt-breadcrumbs-item>
</dt-breadcrumbs>`,
  );

  tree.create(
    'libs/feature-a/src/lib/table-empty-state-simple.html',
    `<dt-table-empty-state>
  <dt-table-empty-state-message>Amend the timeframe you're querying within or review your query to make your statement less restrictive</dt-table-empty-state-message>
</dt-table-empty-state>`,
  );

  tree.create(
    'libs/feature-a/src/lib/table-empty-state.html',
    `<div class="empty-state-container" *ngIf="queryRows?.length === 0">
    <dt-table-empty-state dtTableEmptyState uitestid="dtaqlTableNoData">
      <dt-table-empty-state-image class="no-data-image"></dt-table-empty-state-image>
      <dt-table-empty-state-title i18n>No data that matches your query</dt-table-empty-state-title>
      <dt-table-empty-state-message i18n *ngIf="(filtersApplied$ | async)">Amend the timeframe you're querying within or review your query to make your statement less restrictive</dt-table-empty-state-message>
    </dt-table-empty-state>
  </div>`,
  );

  tree.create(
    'libs/feature-a/src/lib/secondary-nav.html',
    `<dt-secondary-nav i18n-aria-label aria-label="Deployment status"></dt-secondary-nav>`,
  );

  tree.create(
    'libs/feature-a/src/lib/secondary-nav-aria-label.html',
    `<dt-secondary-nav aria-label="Deployment status"></dt-secondary-nav>`,
  );

  tree.create(
    'libs/feature-a/src/lib/aria-label-binding.html',
    `<dt-secondary-nav [attr.aria-label]="'Deployment status'"></dt-secondary-nav>
     <dt-secondary-nav [attr.aria-label]="somevar"></dt-secondary-nav>`,
  );

  return tree;
}
