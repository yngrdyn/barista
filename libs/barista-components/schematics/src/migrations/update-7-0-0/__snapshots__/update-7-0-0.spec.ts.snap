// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Update 7.0.0 Template replacements should migrate aria label bindings 1`] = `
"<dt-secondary-nav [attr.aria-label]=\\"'Deployment status'\\"></dt-secondary-nav>
     <dt-secondary-nav [attr.aria-label]=\\"somevar\\"></dt-secondary-nav>"
`;

exports[`Update 7.0.0 Template replacements should migrate the breadcrumbs-item selector to a plain anchor 1`] = `
"<dt-breadcrumbs aria-label=\\"Breadcrumbs navigation\\" i18n-aria-label>
  <a dtBreadcrumbsItem i18n>Security</a>
  <a dtBreadcrumbsItem href=\\"/home\\" i18n>Security 2</a>
  <a dtBreadcrumbsItem [href]=\\"'/home'\\" i18n>Security 3</a>
</dt-breadcrumbs>"
`;

exports[`Update 7.0.0 Template replacements should migrate the table empty state to the empty state component 1`] = `
"<dt-empty-state><dt-empty-state-item>

  Amend the timeframe you're querying within or review your query to make your statement less restrictive

</dt-empty-state-item></dt-empty-state>"
`;

exports[`Update 7.0.0 Template replacements should migrate the table empty state to the empty state component with directive on the message 1`] = `
"<div class=\\"empty-state-container\\" *ngIf=\\"queryRows?.length === 0\\">
    <dt-empty-state dtTableEmptyState uitestid=\\"dtaqlTableNoData\\"><dt-empty-state-item>

      <dt-empty-state-item-img class=\\"no-data-image\\"></dt-empty-state-item-img>
      <dt-empty-state-item-title i18n>No data that matches your query</dt-empty-state-item-title>
      <ng-container i18n *ngIf=\\"(filtersApplied$ | async)\\">Amend the timeframe you're querying within or review your query to make your statement less restrictive</ng-container>
    
</dt-empty-state-item></dt-empty-state>
  </div>"
`;

exports[`Update 7.0.0 Template replacements should not migrate i18n aria label attributes 1`] = `"<dt-secondary-nav i18n-aria-label aria-label=\\"Deployment status\\"></dt-secondary-nav>"`;

exports[`Update 7.0.0 Template replacements should not migrate normal aria label attributes 1`] = `"<dt-secondary-nav aria-label=\\"Deployment status\\"></dt-secondary-nav>"`;

exports[`Update 7.0.0 Template replacements should replace the aria-label to the native one on the breadcrumbs 1`] = `
"<div *pageHeader  dtTheme='blue'>
    <dt-breadcrumbs aria-label=\\"Breadcrumbs navigation\\" i18n-aria-label>
      <a dtBreadcrumbsItem [navLink]=\\"cmcDsService.getHome()\\" i18n>Home</a>
      <a dtBreadcrumbsItem [navLink]=\\"cmcDsService.getDeploymentStatusMainPage()\\" i18n>Deployment status</a>
      <a dtBreadcrumbsItem [navLink]=\\"cmcDsService.getActiveGateOverview()\\" i18n>ActiveGates</a>
    </dt-breadcrumbs>
  </div>"
`;

exports[`Update 7.0.0 Template replacements should replace the aria-label-close-button on the context dialog 1`] = `
"<dt-context-dialog #contextDialog
    aria-label=\\"Show dashboard modification options\\"
    ariaLabelClose=\\"Hide dashboard modification options\\"
    i18n-aria-label
>"
`;

exports[`Update 7.0.0 Template replacements should run the migration for the aria-label-close 1`] = `
"
  <dt-chart [options]=\\"options\\" [series]=\\"series$ | async\\">
    <dt-chart-timestamp
      ariaLabelClose=\\"Close the selection\\"
      (closed)=\\"closed()\\"
      (valueChanges)=\\"valueChanges($event)\\"
    ></dt-chart-timestamp>
  </dt-chart>

  "
`;
